torchvision/prototype/models/depth/stereo/raft_stereo.py

```diff
@@ -0,0 +1,746 @@
+        if with_block:
+            block_layer = block(in_channels, in_channels, norm_layer=nn.BatchNorm2d, stride=1)
+            return nn.Sequential(block_layer, conv_layer)
+        else:
+            return conv_layer
+
+    def _make_downsampler(self, block, in_channels, out_channels):
+        block1 = block(in_channels, out_channels, norm_layer=nn.BatchNorm2d, stride=2)
+        block2 = block(out_channels, out_channels, norm_layer=nn.BatchNorm2d, stride=1)
+        return nn.Sequential(block1, block2)
+
+    def forward(self, x: Tensor) -> List[Tensor]:
+        x = self.base_encoder(x)
+        outs = [torch.cat([self.out_hidden_states[0](x), self.out_contexts[0](x)], dim=1)]
+        for i in range(0, self.num_level - 1):
+            x = self.downsamplers[i](x)
+            outs.append(torch.cat([self.out_hidden_states[i + 1](x), self.out_contexts[i + 1](x)], dim=1))
+        return outs
+
+
+class ConvGRU(raft.ConvGRU):
+    """Convolutional Gru unit."""
+
+    # Modified from raft.ConvGRU to accept pre-convolved contexts,
+    # see: https://github.com/princeton-vl/RAFT-Stereo/blob/main/core/update.py#L23
+    def forward(self, h, context, x):
+        hx = torch.cat([h, x], dim=1)
+        cz, cr, cq = context
+        z = torch.sigmoid(self.convz(hx) + cz)
+        r = torch.sigmoid(self.convr(hx) + cr)
+        q = torch.tanh(self.convq(torch.cat([r * h, x], dim=1)) + cq)
+        h = (1 - z) * h + z * q
+        return h
+
+
+class MultiLevelUpdateBlock(nn.Module):
+    """The update block which contains the motion encoder and grus
+
+    It must expose a ``hidden_dims`` attribute which is the hidden dimension size of its gru blocks
+    """
+
+    def __init__(self, *, motion_encoder: MotionEncoder, hidden_dims: List[int] = [128, 128, 128]):
+        super().__init__()
+        self.motion_encoder = motion_encoder
+
+        # The GRU input size is the size of previous level hidden_dim plus next level hidden_dim
+        # if this is the first gru, then we replace previous level with motion_encoder output channels
+        # for the last GRU, we dont add the next level hidden_dim
+        gru_input_dims = []
+        for i in range(len(hidden_dims)):
+            input_dim = hidden_dims[i - 1] if i > 0 else motion_encoder.out_channels
+            if i < len(hidden_dims) - 1:
+                input_dim += hidden_dims[i + 1]
+            gru_input_dims.append(input_dim)
+
+        self.grus = nn.ModuleList(
+            [
+                ConvGRU(input_size=gru_input_dims[i], hidden_size=hidden_dims[i], kernel_size=3, padding=1)
+                for i in range(len(hidden_dims))
+            ]
+        )
+
+        self.hidden_dims = hidden_dims
+
+    def _downsample2x(self, x):
+        return F.avg_pool2d(x, 3, stride=2, padding=1)
+
+    def _upsample2x(self, x):
+        _, _, h, w = x.shape
+        new_h, new_w = h * 2, w * 2
+        return F.interpolate(x, size=(new_h, new_w), mode="bilinear", align_corners=True)
```

Comment:
Nit: we can do without these private methods as you use them only once.

Reply:
In my opinion, using private method here actually help the readability. 
```
features = torch.cat([features, self._upsample2x(hidden_states[i + 1])], dim=1)
```
It is clear that we want to do upsample (from the function name) on hidden_states[i+1] (in 1 line). And reader who want to know more detail on how we do the upsampling can read the function.

Do you think it is okay to keep the private function?

Reply:
I agreed that readability can be quite subjective. There is an argument to be made about keeping the styling of the code similar across the code-base. Having 1-liner private methods used once is probably something we try to avoid across the code-base. As far as I see, you already have comments to discuss what the 2 function calls are doing. Do the separate methods really add more?